File: tests/XFAIL/cross_file_func_vars/funcs.rs
use std::cell::{RefCell};
use std::rc::{Rc};

/// Regular function for comparison
pub fn regular_double(x: Rc<RefCell<Option<i32>>>) -> Rc<RefCell<Option<i32>>> {

    return {
            let __tmp_x = (*x.borrow_mut().as_mut().unwrap());
            let __tmp_y = 2;
            Rc::new(RefCell::new(Some(__tmp_x * __tmp_y)))
        };
}

/// Function that returns a function
pub fn make_multiplier(factor: Rc<RefCell<Option<i32>>>) -> Rc<RefCell<Option<Box<dyn Fn(Rc<RefCell<Option<i32>>>) -> Rc<RefCell<Option<i32>>>>>>> {

    return Rc::new(RefCell::new(Some(Box::new(move |x: Rc<RefCell<Option<i32>>>| -> Rc<RefCell<Option<i32>>> {
        return {
            let __tmp_x = (*x.borrow_mut().as_mut().unwrap());
            let __tmp_y = (*factor.borrow_mut().as_mut().unwrap());
            Rc::new(RefCell::new(Some(__tmp_x * __tmp_y)))
        };
    }) as Box<dyn Fn(Rc<RefCell<Option<i32>>>) -> Rc<RefCell<Option<i32>>>>)));
}

pub fn init() {
        // Assign function to variable in init
    { let new_val = Rc::new(RefCell::new(Some(Box::new(move |s: Rc<RefCell<Option<String>>>| -> Rc<RefCell<Option<String>>> {
        return Rc::new(RefCell::new(Some(format!("Dynamic: {}", (*s.borrow_mut().as_mut().unwrap())))));
    }) as Box<dyn Fn(Rc<RefCell<Option<String>>>) -> Rc<RefCell<Option<String>>>>))); *DYNAMIC_FUNC.borrow_mut() = Some(new_val); };
}

File: tests/cross_file_func_vars/funcs.go
package main

import "fmt"

// Function variable - transpiler needs to know this is a function, not a regular variable
var ProcessData = func(x int) int {
	return x * 2
}

// Another function variable with different signature
var CombineStrings = func(a, b string) string {
	return a + " " + b
}

// Function variable that takes a function as parameter
var ApplyTwice = func(f func(int) int, x int) int {
	return f(f(x))
}

// Function variable with no parameters
var GetGreeting = func() string {
	return "Hello from function variable!"
}

// Function variable with multiple returns
var DivMod = func(a, b int) (int, int) {
	return a / b, a % b
}

// Regular function for comparison
func RegularDouble(x int) int {
	return x * 2
}

// Function that returns a function
func MakeMultiplier(factor int) func(int) int {
	return func(x int) int {
		return x * factor
	}
}

// Global variable that will be assigned a function later
var DynamicFunc func(string) string

func init() {
	// Assign function to variable in init
	DynamicFunc = func(s string) string {
		return fmt.Sprintf("Dynamic: %s", s)
	}
}
